runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -457,10 +575,13 @@ function LanguageTree:parse(range)
   })
 
   for _, child in pairs(self._children) do
-    child:parse(range)
+    local _, child_finished = child:_parse(range, timeout)
+    if not child_finished then
+      return self._trees, child_finished
+    end
```

Comment:
If there are lots of children then this can also stack the timeout.

```suggestion
    local ctime, _, child_finished = tcall(child._parse, child, range, timeout)
    timeout = math.max(timeout - ctime, 0)
    if timeout == 0 or not child_finished then
      return self._trees, child_finished
    end
```